libnopegl/src/node_gblur.c

```diff
@@ -0,0 +1,543 @@
+#include "rendertarget.h"
+#include "rtt.h"
+#include "topology.h"
+#include "utils.h"
+
+/* GLSL shaders */
+#include "blur_gaussian_vert.h"
+#include "blur_gaussian_frag.h"
+
+#define _CONSTANT_TO_STR(v) #v
+#define CONSTANT_TO_STR(v) _CONSTANT_TO_STR(v)
+
+#define MAX_KERNEL_SIZE 127
+NGLI_STATIC_ASSERT(kernel_size_is_odd, MAX_KERNEL_SIZE & 0x1);
+
+struct direction_block {
+    float direction[2];
+};
+
+struct kernel_block {
+    float weights[2 * MAX_KERNEL_SIZE];
+    int32_t nb_weights;
+};
+
+struct gblur_opts {
+    struct ngl_node *source;
+    struct ngl_node *destination;
+    struct ngl_node *bluriness_node;
+    float bluriness;
+    struct ngl_node *max_kernel_size_node;
+    uint32_t max_kernel_size;
+};
+
+struct gblur_priv {
+    int32_t width;
+    int32_t height;
+    float bluriness;
+    uint32_t max_kernel_size;
+    uint32_t max_radius_size;
+
+    /* Source image */
+    struct image *image;
+    size_t image_rev;
+
+    /* Render the horizontal pass to a temporary destination */
+    struct rendertarget_layout tmp_layout;
+    struct rtt_ctx *tmp;
+
+    /* Render the vertical pass to the destination */
+    int dst_is_resizeable;
+    struct rendertarget_layout dst_layout;
+    struct rtt_ctx *dst_rtt_ctx;
+
+    struct gpu_block direction;
+    struct gpu_block kernel;
+    struct pgcraft *crafter;
+    struct pipeline_compat *pl_blur_h;
+    struct pipeline_compat *pl_blur_v;
+};
+
+#define OFFSET(x) offsetof(struct gblur_opts, x)
+static const struct node_param gblur_params[] = {
+    {"source",            NGLI_PARAM_TYPE_NODE, OFFSET(source),
+                          .node_types=(const uint32_t[]){NGL_NODE_TEXTURE2D, NGLI_NODE_NONE},
+                          .flags=NGLI_PARAM_FLAG_NON_NULL | NGLI_PARAM_FLAG_DOT_DISPLAY_FIELDNAME,
+                          .desc=NGLI_DOCSTRING("source to use for the blur")},
+    {"destination",       NGLI_PARAM_TYPE_NODE, OFFSET(destination),
+                          .node_types=(const uint32_t[]){NGL_NODE_TEXTURE2D, NGLI_NODE_NONE},
+                          .flags=NGLI_PARAM_FLAG_NON_NULL | NGLI_PARAM_FLAG_DOT_DISPLAY_FIELDNAME,
+                          .desc=NGLI_DOCSTRING("destination to use for the blur")},
+    {"bluriness",         NGLI_PARAM_TYPE_F32, OFFSET(bluriness_node), {.f32=0.03f},
+                          .flags=NGLI_PARAM_FLAG_ALLOW_NODE,
+                          .desc=NGLI_DOCSTRING("amount of bluriness in the range [0, 1]")},
+    {"max_kernel_size",   NGLI_PARAM_TYPE_U32, OFFSET(max_kernel_size_node), {.u32=127},
+                          .flags=NGLI_PARAM_FLAG_ALLOW_NODE,
+                          .desc=NGLI_DOCSTRING("maximum kernel size to be used internally, "
+                          "a higher value increase the maximum blur amount at the cost of "
+                          "performance, must be in the range [3;" CONSTANT_TO_STR(MAX_KERNEL_SIZE) "]")},
+    {NULL}
+};
+
+#define O(i) (2 * (i))
+#define W(i) (O(i) + 1)
```

Comment:
nit: it would probably be clearer to have:
```c
#define O(i) (2 * (i))
#define W(i) (2 * (i) + 1)
```
Otherwise you may think of some sort of relationship while it's just offset packing

Reply:
Changed.